src/interactive/userunique.py

```diff
@@ -0,0 +1,108 @@
+import logging
+
+from typing import Dict, Tuple
+
+import discord
+
+from interactive.selection import TimedView
+
+logger = logging.getLogger(__name__)
+
+
+class PromptModal(discord.ui.Modal, title="e-bot"):
+    def __init__(self, content):
+        super().__init__()
+        self.response = discord.ui.TextInput(label=content)
+        self.add_item(self.response)
+
+    async def on_submit(self, interaction: discord.Interaction):
+        self.stop()
+        self.interaction = interaction
+
+
+class UserPrompt(discord.ui.View):
+    def __init__(self, content, default_outcome: str, **kwargs):
+        super().__init__(**kwargs)
+        self.content = content
+        self.outcome: str | None = None
+        self.default = default_outcome
+        self.used_default = False
+
+    @discord.ui.button(label="Safety")
+    async def safety(self, interaction: discord.Interaction, button):
+        self.outcome = None
+        await self.resolve(interaction.message, interaction)
+
+    @discord.ui.button(label="Enter text")
+    async def user_input(self, interaction: discord.Interaction, button):
+        modal = PromptModal(self.content)
+        await interaction.response.send_modal(modal)
+        # wait for response
+        await modal.wait()
+
+        self.outcome = modal.response.value
+        await self.resolve(interaction.message, modal.interaction)
+
+    async def on_timeout(self):
```

Comment:
What is the behaviour here if the component times out but one of the `await` statements (e.g `await modal.wait()`) is still executing?

Reply:
I believe the discord library kills any hanging coros in that event loop, calls `on_timeout`, and the yields to wherever `await thing.wait()` is?